R/schema.R

```diff
@@ -0,0 +1,301 @@
+#'
+#' @inheritParams sql_schema_exists
+#'
+#' @details Methods are not available for all DBMS.
+#'
+#' An error is thrown if a schema of that name already exists.
+#'
+#' Additional arguments are:
+#'
+#'   - `dbname`: supported for MSSQL. Create a schema in a different
+#'   database on the connected MSSQL-server; default: database addressed by `dest`.
+#'
+#' @return `NULL` invisibly.
+#'
+#' @family schema handling functions
+#' @export
+sql_schema_create <- function(dest, schema, ...) {
+  check_param_class(schema, "character")
+  check_param_length(schema)
+  if (sql_schema_exists(dest, schema, ...)) {abort_schema_exists(schema, ...)}
+  UseMethod("sql_schema_create")
+}
+
+#' @export
+sql_schema_create.src_dbi <- function(dest, schema, ...) {
+  sql_schema_create(dest$con, schema, ...)
+}
+
+#' @export
+sql_schema_create.PqConnection <- function(dest, schema, ...) {
+  DBI::dbExecute(dest, glue::glue("CREATE SCHEMA {schema}"))
+  message(glue::glue("Schema {tick(schema)} created."))
+  invisible(NULL)
+}
+
+#' @export
+`sql_schema_create.Microsoft SQL Server` <- function(dest, schema, dbname = NULL, ...) {
+  if (!is_null(dbname)) {
+    original_dbname <- attributes(dest)$info$dbname
+    DBI::dbExecute(dest, glue::glue("USE {dbname}"))
+    withr::defer(DBI::dbExecute(dest, glue::glue("USE {original_dbname}")))
+    msg_suffix <- paste0(" on database ", tick(dbname))
+  } else {
+    msg_suffix <- ""
+  }
+  DBI::dbExecute(dest, glue::glue("CREATE SCHEMA {schema}"))
+  message(glue::glue("Schema {tick(schema)} created{msg_suffix}."))
+  invisible(NULL)
+}
+
+# sql_schema_table_list() -------------------------------------------------
+
+#' List the tables in a schema on a database
+#'
+#' @description `sql_schema_table_list()` list the tables in a schema on the database.
+#'
+#' @inheritParams sql_schema_exists
+#'
+#' @details Methods are not available for all DBMS.
+#'
+#' An error is thrown if no schema of that name exists.
+#'
+#' Additional arguments are:
+#'
+#'   - `dbname`: supported for MSSQL. Look for tables on a different
+#'   database on the connected MSSQL-server; default: database addressed by `dest`.
+#'
+#' @return A tibble with the following columns:
+#'   \describe{
+#'     \item{`table_name`}{name of the table,}
+#'     \item{`remote_name`}{identifier of the table on the DBMS.
+#'     Can be used to access the listed tables with the syntax
+#'     `tbl(dest, remote_name).`}
+#'   }
+#'
+#' @family schema handling functions
+#' @export
+sql_schema_table_list <- function(dest, schema = NULL, ...) {
+  if (!is_null(schema)) {
+    check_param_class(schema, "character")
+    check_param_length(schema)
+  }
+  if (!is_null(schema) && !sql_schema_exists(dest, schema, ...)) {abort_no_schema_exists(schema, ...)}
+  UseMethod("sql_schema_table_list")
+}
+
+#' @export
+`sql_schema_table_list.src_Microsoft SQL Server` <- function(dest, schema = NULL, dbname = NULL, ...) {
```

Comment:
Do we need the `src_` methods? Either way, we should implement methods without `src_` and if necessary forward the `src_` methods.

Also, listing tables (in a schema) should maybe be part of dbplyr?

Reply:
> [...] we should implement methods without src_ and if necessary forward the src_ methods.

I did it this way with the exception of `sql_schema_table_list()`, where I implemented the `src_`-methods, since we had a handy function for this purpose readily available. I think it's just slightly more comfortable for users not to have to worry, if it's a `src` or a `con` object.

Shall we make a PR for listing the tables in dbplyr? Is the output table something you consider useful? (tibble with 2 columns: table_name and remote_name). In dbplyr, should it be rather a `src`-method than a `con`-method?

Reply:
So I will not export `sql_schema_table_list()` and just use 2 internal functions for now, `sql_schema_table_list_mssql()` and `sql_schema_table_list_postgres()`